runners/apex/src/main/java/org/apache/beam/runners/apex/ApexRunner.java

```diff
@@ -0,0 +1,166 @@
+import org.apache.beam.sdk.util.WindowingStrategy;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.hadoop.conf.Configuration;
+
+import com.datatorrent.api.Context.DAGContext;
+import com.datatorrent.api.DAG;
+import com.datatorrent.api.LocalMode;
+import com.datatorrent.api.StreamingApplication;
+import com.google.common.base.Throwables;
+
+/**
+ * A {@link PipelineRunner} that translates the
+ * pipeline to an Apex DAG and executes it on an Apex cluster.
+ * <p>
+ * Currently execution is always in embedded mode,
+ * launch on Hadoop cluster will be added in subsequent iteration.
+ */
+@SuppressWarnings({"rawtypes", "unchecked"})
+public class ApexRunner extends PipelineRunner<ApexRunnerResult> {
+
+  private final ApexPipelineOptions options;
+
+  public ApexRunner(ApexPipelineOptions options) {
+    this.options = options;
+  }
+
+  public static ApexRunner fromOptions(PipelineOptions options) {
+    return new ApexRunner((ApexPipelineOptions) options);
+  }
+
+  @Override
+  public <OutputT extends POutput, InputT extends PInput> OutputT apply(
+      PTransform<InputT, OutputT> transform, InputT input) {
+    if (Window.Bound.class.equals(transform.getClass())) {
+      return (OutputT) ((PCollection) input).apply(
+          new AssignWindowsAndSetStrategy((Window.Bound) transform));
+    } else if (Create.Values.class.equals(transform.getClass())) {
+      return (OutputT) PCollection
+          .<OutputT>createPrimitiveOutputInternal(
+              input.getPipeline(),
+              WindowingStrategy.globalDefault(),
+              PCollection.IsBounded.BOUNDED);
+    } else {
+      return super.apply(transform, input);
+    }
+  }
+
+  @Override
+  public ApexRunnerResult run(Pipeline pipeline) {
+
+    final TranslationContext translationContext = new TranslationContext(options);
+    ApexPipelineTranslator translator = new ApexPipelineTranslator(translationContext);
+    translator.translate(pipeline);
+
+    StreamingApplication apexApp = new StreamingApplication()
+    {
+      @Override
+      public void populateDAG(DAG dag, Configuration conf)
+      {
+        dag.setAttribute(DAGContext.APPLICATION_NAME, options.getApplicationName());
+        translationContext.populateDAG(dag);
+      }
+    };
+
+    checkArgument(options.isEmbeddedExecution(), "only embedded execution is supported at this time");
+    LocalMode lma = LocalMode.newInstance();
+    Configuration conf = new Configuration(false);
+    try {
+      lma.prepareDAG(apexApp, conf);
+      LocalMode.Controller lc = lma.getController();
+      if (options.isEmbeddedExecutionDebugMode()) {
+        // turns off timeout checking for operator progress
+        lc.setHeartbeatMonitoringEnabled(false);
+      }
+      if (options.getRunMillis() > 0) {
+        lc.run(options.getRunMillis());
+      } else {
+        lc.runAsync();
```

Comment:
There has been some discussion of this on the mailing list. The consensus is that the simple way to go is to always `runAsync` and then `PipelineResult#waitUntilFinish()` can be called if someone wants a synchronous call.

For existing runners, there are some complications remaining, since for testing you really want to call `waitToFinish()` so all assertions run and throw exceptions, etc. Eventually this will probably be handled by `TestPipeline.run()` but in the meantime you can make a `TestApexRunner` that does `result = apexRunner.run(); result.waitToFinish()`.